scm-record/src/bin/scm-diff-editor.rs

```diff
@@ -0,0 +1,606 @@
+                            lines.push(line);
+                        }
+                        _ => {
+                            acc.push(Section::Changed { lines: vec![line] });
+                        }
+                    }
+                }
+                diffy::Line::Insert(line) => {
+                    let line = SectionChangedLine {
+                        is_toggled: false,
+                        change_type: ChangeType::Added,
+                        line: Cow::Owned((*line).to_owned()),
+                    };
+                    match acc.last_mut() {
+                        Some(Section::Changed { lines }) => {
+                            lines.push(line);
+                        }
+                        _ => {
+                            acc.push(Section::Changed { lines: vec![line] });
+                        }
+                    }
+                }
+            }
+            acc
+        }));
+    }
+    sections
+}
+
+fn print_dry_run(write_root: &Path, state: RecordState) {
+    let scm_record::RecordState { files } = state;
+    for file in files {
+        let file_path = write_root.join(file.path.clone());
+        let (selected_contents, _unselected_contents) = file.get_selected_contents();
+        match selected_contents {
+            scm_record::SelectedContents::Absent => {
+                println!("Would delete file: {}", file_path.display())
+            }
+            scm_record::SelectedContents::Unchanged => {
+                println!("Would leave file unchanged: {}", file_path.display())
+            }
+            scm_record::SelectedContents::Binary {
+                old_description,
+                new_description,
+            } => {
+                println!("Would update binary file: {}", file_path.display());
+                println!("  Old: {:?}", old_description);
+                println!("  New: {:?}", new_description);
+            }
+            scm_record::SelectedContents::Present { contents } => {
+                println!("Would update text file: {}", file_path.display());
+                for line in contents.lines() {
+                    println!("  {line}");
+                }
+            }
+        }
+    }
+}
+
+fn apply_changes(write_root: &Path, state: RecordState) -> Result<(), Error> {
+    let scm_record::RecordState { files } = state;
+    for file in files {
+        let file_path = write_root.join(file.path.clone());
+        let (selected_contents, _unselected_contents) = file.get_selected_contents();
+        match selected_contents {
+            scm_record::SelectedContents::Absent => match fs::remove_file(&file_path) {
+                Ok(()) => {}
+                Err(err) if err.kind() == io::ErrorKind::NotFound => {}
+                Err(err) => {
+                    return Err(Error::RemoveFile {
+                        path: file_path,
+                        source: err,
+                    });
+                }
+            },
```

Comment:
For added files that have not been selected, I guess `get_selected_contents()` will not mention the file, so we won't get an `Absent` match here even though we want to remove the file.